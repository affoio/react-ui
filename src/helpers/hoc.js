import { createElement } from 'react';
import { omit } from 'ramda';

const componentFromProp = propName => {
  const Component = props => createElement(props[propName], omit([propName], props));
  Component.displayName = `componentFromProp(${propName})`;
  return Component;
};

export default componentFromProp;
