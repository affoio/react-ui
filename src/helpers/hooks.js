import { useEffect, useState } from 'react';

export const useOnClickOutside = (ref, handler) => {
  useEffect(() => {
    const listener = event => {
      // Do nothing if clicking ref's element or descendent elements
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }

      handler(event);
    };

    document.addEventListener('click', listener);
    document.addEventListener('touchstart', listener);
    document.addEventListener('focus', listener, true);

    return () => {
      document.removeEventListener('click', listener);
      document.removeEventListener('touchstart', listener);
      document.removeEventListener('focus', listener);
    };
  }, [ref, handler]);
};

export const useDebounce = (value, delay) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [delay, value]);

  return debouncedValue;
};
