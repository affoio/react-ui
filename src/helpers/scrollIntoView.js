import smoothscroll from 'smoothscroll-polyfill';

smoothscroll.polyfill();

export function scrollIntoView(menuEl, focusedEl) {
  const menuRect = menuEl.getBoundingClientRect();
  const focusedRect = focusedEl.getBoundingClientRect();
  const overScroll = focusedEl.offsetHeight / 3;

  if (focusedRect.bottom + overScroll > menuRect.bottom) {
    menuEl.scrollTo(
      menuEl,
      Math.min(focusedEl.offsetTop + focusedEl.clientHeight - menuEl.offsetHeight + overScroll, menuEl.scrollHeight)
    );
  } else if (focusedRect.top - overScroll < menuRect.top) {
    menuEl.scrollTo(menuEl, Math.max(focusedEl.offsetTop - overScroll, 0));
  }
}
