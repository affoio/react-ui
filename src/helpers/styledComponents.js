import { css } from 'styled-components';

export const addStyleIfPropTruly = (propName, style) => props => {
  if (props[propName]) {
    return css`
      ${style}
    `;
  }
};
