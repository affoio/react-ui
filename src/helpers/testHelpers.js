import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'enzyme';

import { MemoryRouter } from 'react-router';

export const renderWithThemeProvider = (children, theme = {}) =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const withProviders = (component, initialEntries, theme = {}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  </MemoryRouter>
);
