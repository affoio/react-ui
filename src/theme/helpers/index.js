import { css } from 'styled-components';

export const isInteger = value => parseFloat(value) === parseInt(value, 10);
export const calculatePercentageGridValue = (value, gridSize) => `${(100 / gridSize) * value}%`;
export const getLimitedMedia = (breakpoints, current, currentIndex) => {
  return currentIndex !== Object.keys(breakpoints).length - 1
    ? `(min-width: ${breakpoints[current]}px) and (max-width: ${breakpoints[
        Object.keys(breakpoints)[currentIndex + 1]
      ] - 1}px)`
    : `(min-width: ${breakpoints[current]}px)`;
};

export const mediaQueryFunctionsList = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] =
      index === 0
        ? (...args) => css`
            ${css(...args)}
          `
        : (...args) => css`
            @media (min-width: ${breakpoints[breakpoint]}px) {
              ${css(...args)}
            }
          `;
    return newMedia;
  }, {});

export const mediaQueryFunctionsListWithLimits = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] = (...args) => css`
      @media ${getLimitedMedia(breakpoints, breakpoint, index)} {
        ${css(...args)}
      }
    `;
    return newMedia;
  }, {});
