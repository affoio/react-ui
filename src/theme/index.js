export const defaultTheme = {
  baseIndent: 4,
  breakpoints: { xs: 320, sm: 768, lg: 1280 },
  gridSize: 12,
  gutterWidth: 2,
  container: {
    indent: {
      xs: 4,
      sm: 4,
      lg: 15,
    },
    maxWidth: '1920px',
  },
  typography: {
    '100': { fontSize: '12px', lineHeight: '20px' },
    '200': { fontSize: '14px', lineHeight: '23px' },
    '300': { fontSize: '16px', lineHeight: '25px' },
    '400': { fontSize: '18px', lineHeight: '28px' },
    '500': { fontSize: '24px', lineHeight: '32px' },
    '600': { fontSize: '32px', lineHeight: '44px' },
  },
};

export const themeConfig = {
  breakpointNames: ['xs', 'sm', 'lg'],
  breakpointOffsetNames: ['xsOffset', 'smOffset', 'lgOffset'],
};
