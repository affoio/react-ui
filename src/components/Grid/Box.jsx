import React from 'react';
import { number, string } from 'prop-types';
import styled, { css } from 'styled-components';
import { getMultipliedIndent } from '../../theme/selectors';
import { addStyleIfPropTruly } from '../../helpers/styledComponents';

export const propTypes = {
  py: number,
  px: number,
  pb: number,
  pt: number,
  mx: number,
  my: number,
  mb: number,
  mt: number,
  mr: number,
  ml: number,
  maxWidth: string,
  width: string,
};

const Box = styled.div`
  box-sizing: border-box;
  ${({ py, theme }) => {
    if (py) {
      const indent = getMultipliedIndent(theme, py);

      return css`
        padding-top: ${indent};
        padding-bottom: ${indent};
      `;
    }
  }}
  ${({ px, theme }) => {
    if (px) {
      const indent = getMultipliedIndent(theme, px);

      return css`
        padding-left: ${indent};
        padding-right: ${indent};
      `;
    }
  }}
  
  ${({ pb, theme }) => {
    if (pb) {
      const indent = getMultipliedIndent(theme, pb);

      return css`
        padding-bottom: ${indent};
      `;
    }
  }}
  
  ${({ pt, theme }) => {
    if (pt) {
      const indent = getMultipliedIndent(theme, pt);

      return css`
        padding-top: ${indent};
      `;
    }
  }}
  
  ${({ mx, theme }) => {
    if (mx) {
      const indent = getMultipliedIndent(theme, mx);

      return css`
        margin-left: ${indent};
        margin-right: ${indent};
      `;
    }
  }}
  
  ${({ my, theme }) => {
    if (my) {
      const indent = getMultipliedIndent(theme, my);

      return css`
        margin-top: ${indent};
        margin-bottom: ${indent};
      `;
    }
  }}
  
  ${({ mb, theme }) => {
    if (mb) {
      const indent = getMultipliedIndent(theme, mb);

      return css`
        margin-bottom: ${indent};
      `;
    }
  }}
  
  ${({ mt, theme }) => {
    if (mt) {
      const indent = getMultipliedIndent(theme, mt);

      return css`
        margin-top: ${indent};
      `;
    }
  }}
  ${({ mr, theme }) => {
    if (mr) {
      const indent = getMultipliedIndent(theme, mr);

      return css`
        margin-right: ${indent};
      `;
    }
  }}
  
  ${({ ml, theme }) => {
    if (ml) {
      const indent = getMultipliedIndent(theme, ml);

      return css`
        margin-left: ${indent};
      `;
    }
  }}

  ${({ width }) => width && `width: ${width};`}
  ${({ maxWidth }) => maxWidth && `max-width: ${maxWidth};`}
  ${addStyleIfPropTruly('flex', 'display: flex')}
`;

Box.propTypes = propTypes;

export default React.memo(Box);
