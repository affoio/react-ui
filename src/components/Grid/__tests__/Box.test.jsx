import React from 'react';
import renderer from 'react-test-renderer';
import { defaultTheme } from '../../../theme';
import Box from '../Box';
import { renderWithThemeProvider } from '../../../helpers/testHelpers';

describe('<Box>', () => {
  it('renders Box', () => {
    const tree = renderWithThemeProvider(<Box>Test Text</Box>);

    expect(tree).toMatchSnapshot();
  });

  it('add width with width prop', () => {
    const tree = renderer.create(<Box width="100px">Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('width', '100px');
  });

  it('adds max-width if maxWidth', () => {
    const tree = renderer.create(<Box maxWidth="100%">Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('max-width', '100%');
  });

  it('add horizontal paddings with px prop', () => {
    const tree = renderer.create(<Box px={4}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('padding-left', `${4 * defaultTheme.baseIndent}px`);
    expect(tree).toHaveStyleRule('padding-right', `${4 * defaultTheme.baseIndent}px`);
  });

  it('add vertical paddings with py prop', () => {
    const tree = renderer.create(<Box py={4}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('padding-top', `${4 * defaultTheme.baseIndent}px`);
    expect(tree).toHaveStyleRule('padding-bottom', `${4 * defaultTheme.baseIndent}px`);
  });

  it('add padding-top with pt prop', () => {
    const tree = renderer.create(<Box pt={4}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('padding-top', `${4 * defaultTheme.baseIndent}px`);
  });

  it('add bottom padding with pb prop', () => {
    const tree = renderer.create(<Box pb={4}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('padding-bottom', `${4 * defaultTheme.baseIndent}px`);
  });

  it('add horizontal margins with mx prop', () => {
    const tree = renderer.create(<Box mx={2}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('margin-left', `${2 * defaultTheme.baseIndent}px`);
    expect(tree).toHaveStyleRule('margin-right', `${2 * defaultTheme.baseIndent}px`);
  });

  it('add vertical margins with py prop', () => {
    const tree = renderer.create(<Box my={2}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('margin-top', `${2 * defaultTheme.baseIndent}px`);
    expect(tree).toHaveStyleRule('margin-bottom', `${2 * defaultTheme.baseIndent}px`);
  });

  it('add margin-top with mt prop', () => {
    const tree = renderer.create(<Box mt={2}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('margin-top', `${2 * defaultTheme.baseIndent}px`);
  });

  it('add bottom margin with mb prop', () => {
    const tree = renderer.create(<Box mb={2}>Test Text</Box>).toJSON();

    expect(tree).toHaveStyleRule('margin-bottom', `${2 * defaultTheme.baseIndent}px`);
  });
});
