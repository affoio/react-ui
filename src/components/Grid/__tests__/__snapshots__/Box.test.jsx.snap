// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box> renders Box 1`] = `
.c0 {
  box-sizing: border-box;
}

<div
  class="c0"
>
  Test Text
</div>
`;
